### PR TITLE
config: check if current config is the equivalent of last one

### LINKED ISSUE #8438: Last configuration can be corrupted when disk is full
## Bug Report

TiKV save current configuration before startup. If disk is full at that time, the configuration can be partially saved hence corrupted. It's recommended to check if the configuration is the same before saving so that a lot of writes can be saved. If they are different, we should dump and rename instead of writing directly.

### PR DESCRIPTION
**Pull request by gentcys**
Signed-off-by: gentcys <iamchencong@gmail.com>

### What problem does this PR solve?

Issue Number: close #8438 

### What is changed and how it works?

What's Changed:
If current config is the equivalent of last one while startup, use last one directly instead of writing current one to the file.

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: src/config.rs [hunk 1]
```diff
@@ -2425,6 +2425,12 @@ pub fn persist_config(config: &TiKvConfig) -> Result<(), String> {
     let last_cfg_path = store_path.join(LAST_CONFIG_FILE);
     let tmp_cfg_path = store_path.join(TMP_CONFIG_FILE);
 
+    if let Some(last_cfg) = get_last_config(&config.storage.data_dir) {
+        if serde_json::to_string(&config).unwrap() == serde_json::to_string(&last_cfg).unwrap() {
```
Comment (BusyJay) [inline src/config.rs:2429]: How about:
```
let same = fs::read_to_string(&last_cfg_path).map_or(false, |last_cfg| {
    serde_json::to_string(&config).unwrap() == last_cfg
});
if same {
    return Ok(());
}
```

Comment (gentcys) [reply src/config.rs:2429]: @BusyJay Nice! I'd wondered about how to read last config by `tmp_cfg_path`. And it's more readable to declare a   variable `same` first.

File: src/config.rs [hunk 2]
```diff
@@ -2771,6 +2778,23 @@ mod tests {
         assert!(tikv_cfg.check_critical_cfg_with(&last_cfg).is_ok());
     }
 
+    #[test]
+    fn test_last_cfg_not_modified_if_it_is_equivalent() {
```
Comment (BusyJay) [inline src/config.rs:2782]: Should also check updates are actually triggering writes.

**Comment by BusyJay** [comment]
CI fails because creation time is not supported. You may need to use modified time instead.

File: src/config.rs (hunk 2 above)
Comment (BusyJay) [reply src/config.rs:2782]: Any progress on this?

Comment (gentcys) [reply src/config.rs:2782]: I'll add a test about this.

File: src/config.rs [hunk 3]
```diff
@@ -2771,6 +2778,31 @@ mod tests {
         assert!(tikv_cfg.check_critical_cfg_with(&last_cfg).is_ok());
     }
 
+    #[test]
+    fn test_last_cfg_modified() {
+        let mut tikv_cfg = TiKvConfig::default();
```
Comment (NingLin-P) [inline src/config.rs:2783]: Should use `let (mut cfg, tmp_dir) = TiKvConfig::with_tmp().unwrap()` to avoid leaving config files after runing test.

**Review by NingLin-P — COMMENTED** [review]
Rest LGTM

File: src/config.rs [hunk 4]
```diff
@@ -2791,10 +2791,16 @@ mod tests {
         let mut last_cfg_metadata = last_cfg_path.metadata().unwrap();
         let first_modified = last_cfg_metadata.modified().unwrap();
 
+        // not write to file when config is the equivalent of last one.
         assert!(persist_config(&tikv_cfg).is_ok());
-
         last_cfg_metadata = last_cfg_path.metadata().unwrap();
         assert_eq!(last_cfg_metadata.modified().unwrap(), first_modified);
```
Comment (gentcys) [inline src/config.rs:2797]:
```
test config::tests::test_last_cfg_modified ... thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `SystemTime { tv_sec: 1598249476, tv_nsec: 91041229 }`,
 right: `SystemTime { tv_sec: 1598249476, tv_nsec: 82041320 }`', src/config.rs:2797:9
```
Why are two `modified` time are same in seconds but different in nanoseconds?

**Review by gentcys — COMMENTED** [review]
Test failed, am i missing something?

File: src/config.rs (hunk 4 above)
Comment (BusyJay) [reply src/config.rs:2797]: Then `persist_config` is triggering writes unexpectedly.

**Comment by gentcys** [comment]
I make a force git push to my personal branch to remove the commit without adding a `Signed-off-by`.

File: src/config.rs (hunk 3 above)
Comment (gentcys) [reply src/config.rs]: Thanks!

File: src/config.rs (hunk 4 above)
Comment (gentcys) [reply src/config.rs:2795]: I've fixed it.

**Review by NingLin-P — APPROVED** [review]
LGTM, thanks!

**Review by BusyJay — APPROVED** [review]

**Pull request merged**